trainer_tf_barebones.py

```diff
@@ -0,0 +1,366 @@
+"""Tensorflow trainer class."""
+
+import itertools
+import logging
+import math
+import os
+from typing import Optional
+
+import numpy as np
+import tensorflow as tf
+from sklearn.metrics import classification_report
+from tensorflow_core.python.distribute.distribute_lib import Strategy
+
+from data_processors import DatasetInfo
+from tf_training_args import TFTrainingArguments
+from transformers import (
+    AdamWeightDecay,
+    GradientAccumulator,
+    TFPreTrainedModel,
+    WarmUp,
+)
+
+
+logger = logging.getLogger(__name__)
+
+
+class TFTrainer:
+    model: TFPreTrainedModel
+    args: TFTrainingArguments
+    train_dataset: Optional[tf.data.Dataset]
+    eval_dataset: Optional[tf.data.Dataset]
+    test_dataset: Optional[tf.data.Dataset]
+    dataset_info: DatasetInfo
+
+    strategy: Strategy
+
+    def __init__(
+        self,
+        model: TFPreTrainedModel,
+        args: TFTrainingArguments,
+        train_dataset: Optional[tf.data.Dataset] = None,
+        eval_dataset: Optional[tf.data.Dataset] = None,
+        test_dataset: Optional[tf.data.Dataset] = None,
+        dataset_info: Optional[DatasetInfo] = None,
+    ):
+        self.model = model
+        self.args = args
+        self.train_dataset = train_dataset
+        self.eval_dataset = eval_dataset
+        self.test_dataset = test_dataset
+        self.dataset_info = dataset_info
+
+        self.gradient_accumulator = GradientAccumulator()
+        self.accum_steps = 1
+
+        if self.args.strategy_name == "mirrored":
+            self.strategy = tf.distribute.MirroredStrategy()
+        elif self.args.strategy_name == "onedevice":
+            if len(tf.config.list_physical_devices('GPU')) >= 1:
+                self.strategy = tf.distribute.OneDeviceStrategy(device="/gpu:0")
+            else:
+                self.strategy = tf.distribute.OneDeviceStrategy(device="/cpu:0")
+        else:
+            raise ValueError("The strategy {} does not exists.".format(self.args.strategy_name))
+
+        # To conform with Trainer's API we call this from here.
+        # All args should be in the `args` already.
+        self._setup_training()
```

Comment:
Ok, but I suppose that the `checkpoint_path` and `log_path` parameters will be defined in the `TrainingArguments` and then passed to it?

Reply:
For checkpoint_path, in PyTorch we just hardcode that it's going to be `{args.output_dir}/checkpoint-{global_step}`, cf. https://github.com/huggingface/transformers/blob/6af5a54c28c84e248dac2df2ad7b40968fded3da/src/transformers/trainer.py#L361

For log_path, yes, it's called logging_dir: https://github.com/huggingface/transformers/blob/6af5a54c28c84e248dac2df2ad7b40968fded3da/src/transformers/training_args.py#L60